### PR TITLE
Improve README.md

### PR DESCRIPTION
**Pull request by qurbonzoda**
The changes were developed within GSoC 2023 program by our mentee @wldeh. 
To facilitate review process: 
* The overhauled README focuses on setting up a Kotlin Multiplatform project for writing benchmarks and configuring their execution.
* There will be separate guides on setting up a Kotlin/JVM project or adding a benchmark module in an existing project.
* The non-existing documents referenced in this guide will be added in the followup PRs
* The README describes the next release (0.4.10) with https://github.com/Kotlin/kotlinx-benchmark/pull/163 , https://github.com/Kotlin/kotlinx-benchmark/pull/164 , https://github.com/Kotlin/kotlinx-benchmark/pull/165 , https://github.com/Kotlin/kotlinx-benchmark/pull/166 included

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,273 +1,440 @@
...
+
+        @TearDown
+        fun cleanup() {
+            list.clear()
+        }
+    }
+
+Note: Benchmark classes located in the common source set will be run in all platforms, while those located in a platform-specific source set will be run only in the corresponding platform.
+
+See [writing benchmarks](docs/writing-benchmarks.md) for a complete guide for writing benchmarks.
```
Comment (fzhinkin) [inline README.md:391]: `docs` are not here yet, so it is probably worth removing the reference and adding it back along with other documentation.

File: README.md [hunk 2]
```diff
@@ -1,273 +1,440 @@
...
 
-Register it by its source set name (`jvmBenchmark` is the name for the `benchmark` compilation for `jvm` target):
+### Running Benchmarks
 
-```groovy
+To run your benchmarks in all registered platforms, run `benchmark` Gradle task in your project.
+To run only on a specific platform, run `<target-name>Benchmark`, e.g., `jvmBenchmark`.
+
+For more details about the tasks created by the kotlinx-benchmark plugin, refer to [this guide](docs/tasks-overview.md).
```
Comment (fzhinkin) [inline README.md:398]: The same issue with the link.

File: README.md [hunk 3]
```diff
@@ -1,273 +1,440 @@
...
+            iterations = 3
+            iterationTime = 500
+            iterationTimeUnit = "ms"
+        }
     }
 }
-```
+```  
+
+Refer to our [comprehensive guide](docs/configuration-options.md) to learn about configuration options and how they affect benchmark execution.
```
Comment (fzhinkin) [inline README.md:426]: And here, the link is being dangling as well.

File: README.md [hunk 4]
```diff
@@ -1,273 +1,440 @@
...
+
+### Separate source sets for benchmarks
+
+Often you want to have benchmarks in the same project, but separated from main code, much like tests.
+Refer to our [detailed documentation](docs/separate-source-sets.md) on configuring your project to add a separate source set for benchmarks.
```
Comment (fzhinkin) [inline README.md:431]: Not sure if we can merge it without either adding all new docs from `docs`, or leaving the old description.

**Review by fzhinkin — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -1,273 +1,440 @@
...
-    }   
-}
-```
+1.  **Applying Benchmark Plugin**: Apply the benchmark plugin.
 
-For Kotlin/JVM code, add `allopen` plugin to make JMH happy. Alternatively, make all benchmark classes and methods `open`.
+    ```kotlin
+    // build.gradle.kts
+    plugins {
+        id("org.jetbrains.kotlinx.benchmark") version "0.4.10"
```
[the remark this answers is not in the document]
Comment (qurbonzoda) [reply README.md:66]: As mentioned in the PR description, these changes describe the next release (0.4.10)
I will merge this PR after merging https://github.com/Kotlin/kotlinx-benchmark/pull/163 , https://github.com/Kotlin/kotlinx-benchmark/pull/164 , https://github.com/Kotlin/kotlinx-benchmark/pull/165 , https://github.com/Kotlin/kotlinx-benchmark/pull/166 and releasing a new version.

File: README.md [hunk 6]
```diff
@@ -1,273 +1,440 @@
...
+2. **Set up Parameters and Variables**: Define variables needed for the benchmark.
 
-```groovy
-benchmark {
-    targets {
-        register("benchmarks")    
+    ```kotlin
+    private val size = 10
+
+    private val list = ArrayList()
```
Comment (ilya-g) [inline README.md:334]: type argument is missing

File: README.md [hunk 7]
```diff
@@ -1,273 +1,440 @@
...
+    ```
+
+3. **Initialize Resources**: Within the class, you can define any setup or teardown methods using `@Setup` and `@TearDown` annotations respectively. These methods will be executed before and after the entire benchmark run.
+
+    ```kotlin
+    @Setup
+    fun prepare() {
+        for (i in 0 until size) {
```
Comment (ilya-g) [inline README.md:342]: Minor: `..<` instead of `until`

File: README.md [hunk 8]
```diff
@@ -1,273 +1,440 @@
...
-[![Build status](https://teamcity.jetbrains.com/guestAuth/app/rest/builds/buildType:(id:KotlinTools_KotlinxCollectionsImmutable_Build_All)/statusIcon.svg)](https://teamcity.jetbrains.com/viewType.html?buildTypeId=KotlinTools_KotlinxBenchmark_Build_All)
+[![Build status](https://teamcity.jetbrains.com/guestAuth/app/rest/builds/buildType:(id:KotlinTools_KotlinxBenchmark_Build_All)/statusIcon.svg)](https://teamcity.jetbrains.com/viewType.html?buildTypeId=KotlinTools_KotlinxBenchmark_Build_All)
 [![Maven Central](https://img.shields.io/maven-central/v/org.jetbrains.kotlinx/kotlinx-benchmark-runtime.svg?label=Maven%20Central)](https://search.maven.org/search?q=g:%22org.jetbrains.kotlinx%22%20AND%20a:%22kotlinx-benchmark-runtime%22)
 [![Gradle Plugin Portal](https://img.shields.io/maven-metadata/v?label=Gradle%20Plugin&metadataUrl=https://plugins.gradle.org/m2/org/jetbrains/kotlinx/kotlinx-benchmark-plugin/maven-metadata.xml)](https://plugins.gradle.org/plugin/org.jetbrains.kotlinx.benchmark)
 [![IR](https://img.shields.io/badge/Kotlin%2FJS-IR%20supported-yellow)](https://kotl.in/jsirsupported)
```
Comment (ilya-g) [inline README.md:9]: This badge (JS: IR supported) is likely obsolete.

File: README.md (hunk 7 above)
Comment (Krobysin) [reply README.md:342]: clown

File: README.md (hunk 6 above)
Comment (Krobysin) [reply README.md:334]: clown

File: README.md (hunk 8 above)
Comment (qurbonzoda) [reply README.md]: Removed

File: README.md (hunk 7 above)
Comment (qurbonzoda) [reply README.md]: Replaced

File: README.md (hunk 6 above)
Comment (qurbonzoda) [reply README.md]: Added

File: README.md (hunk 4 above)
Comment (qurbonzoda) [reply README.md:431]: This documentation is added in a separate PR: https://github.com/Kotlin/kotlinx-benchmark/pull/175

File: README.md (hunk 3 above)
Comment (qurbonzoda) [reply README.md:426]: The document is added in a followup PR: https://github.com/Kotlin/kotlinx-benchmark/pull/168

File: README.md (hunk 2 above)
Comment (qurbonzoda) [reply README.md:398]: The document is added in a followup PR: https://github.com/Kotlin/kotlinx-benchmark/pull/170

File: README.md (hunk 1 above)
Comment (qurbonzoda) [reply README.md:389]: The document is added in a followup PR: https://github.com/Kotlin/kotlinx-benchmark/pull/184

**Review by fzhinkin — APPROVED** [review]

**Pull request merged**